enhancements/machine-api/defaulting-validating.md

```diff
@@ -0,0 +1,137 @@
+We could implement `webhook.Defaulter` and `webhook.Validator` interfaces as in https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html.
+
+However we need a more granular level of customisation which is not satisfied by those interfaces. Particularly we need to know the provider we are running on (to run the appropriate defaulting/validation) and the clusterID (to satisfy some defaults).
+To that end we can implement the [Handler interface](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/webhook/admission#Handler) enabling us to infer the values above at the time of instantiating the webhook rather than on each request at runtime.
+
+### Risks and Mitigations
+
+As we introduce stronger opinions on defaults and validation we might break automated consumers that were relying on faulty behaviour.
+New defaults and validation needs to be reviewed carefully for each particular resource.
+
+## Design Details
+
+### Test Plan
+
+Extensive unit testing to cover defaulting and validation business logic.
+e2e testing on the Machine API test suite to validate we reject invalid input.
+
+### Graduation Criteria
+
+
+### Upgrade / Downgrade Strategy
+
+The webhook will start gating upcoming machine API requests as it runs with new payload.
+Existing functional machines should remain untouched. Either because we choose to disable defaulting for update requests or because our defaults should not apply to non-required fields.
+
+### Version Skew Strategy
+
+
+## Implementation History
+
+[AWS implementation](https://github.com/openshift/machine-api-operator/pull/601)
+
+## Drawbacks
+
+
+## Alternatives
+
+- Introduce and new CRD for the `providerSpec`. Externalise it from the machine object. Introduce an `objectRef` in the machine object to reference `providerSpec` resources. This would let the `providerSpec` become a natural extension of the Kubernetes API. Although we still might need a webhook for clever defaulting/validation we'd get for free all the features coming from [structural schemes](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190425-structural-openapi.md)
+
+The above is still desirable but more of an orthogonal long term goal. Once we go that route it means API breaking changes will be disruptive for users.
+
+In order to improve the current UX with defaults and validation, this alternative is not a requirement but a complement.
+
+- Instead of running the webhook server from one of the cross provider controllers i.e machineSet we could provide some facility library and let each particular provider to manage it. This split would increase complexity and would only let us enable the `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` once every single provider implementation have udpated their machine controller to run the webhook server to satisfy the `Endpoint` for the `Service`.
+
```

Comment:
Could MAO not detect the platform and install the MWC/VWC only if the platform supports it?

Reply:
yeh it could, but that'd mean we put that burden on mao when we can get it from cvo for free. It'd also mean we increase design skew between platforms resulting in possible side effects. Also It'd mean we need to coordinate with revendor across multiple repos and teams.
I think having it all coordinated in one single place buys a lot and removes a lot of the headaches while keeping the same level of independence for providers/teams to implement their defaulting/validation creating a PR against mao.